File: backend/src/test/java/com/skewwhiffy/auraltester/controller/ClefControllerTest.java
package com.skewwhiffy.auraltester.controller;

import com.skewwhiffy.auraltester.dto.ClefResponse;
import com.skewwhiffy.auraltester.notation.factory.InternalNotationFactory;
import com.skewwhiffy.auraltester.notation.model.clef.Clef;
import com.skewwhiffy.auraltester.service.AbcService;
import com.skewwhiffy.auraltester.test.util.TestData;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ClefControllerTest {
    @Mock
    private InternalNotationFactory internalNotationFactory;

    @Mock
    private AbcService abcService;

    @InjectMocks
    private ClefController clefController;
    private String clefString;
    private Clef clef;
    private String abc;

    @BeforeEach
    void setUp() {
        clefString = TestData.random().string();
        clef = TestData.random().clef();
        abc = TestData.random().string();
    }

    @Test
    void getsClefInformation() {
        val expected = new ClefResponse(abc);
        when(internalNotationFactory.clef(clefString)).thenReturn(clef);
        when(abcService.getAbc(clef)).thenReturn(() -> abc);

        val actual = clefController.get(clefString);

        assertThat(actual).isEqualTo(expected);
    }
}

File: backend/src/test/kotlin/com/skewwhiffy/auraltester/controller/ClefControllerTest.kt
package com.skewwhiffy.auraltester.controller

import com.ninjasquad.springmockk.MockkBean
import com.skewwhiffy.auraltester.dto.ClefResponse
import com.skewwhiffy.auraltester.notation.factory.InternalNotationFactory
import com.skewwhiffy.auraltester.notation.model.abc.AbcProvider
import com.skewwhiffy.auraltester.notation.model.clef.Clef
import com.skewwhiffy.auraltester.service.AbcService
import com.skewwhiffy.auraltester.test.util.TestData
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class ClefControllerTest {
    @MockkBean
    private lateinit var internalNotationFactory: InternalNotationFactory

    @MockkBean
    private lateinit var abcService: AbcService

    @Autowired
    private lateinit var clefController: ClefController

    private lateinit var clefString: String
    private lateinit var clef: Clef
    private lateinit var abc: String

    @BeforeEach
    fun setUp() {
        clefString = TestData.random.string
        clef = TestData.random.clef
        abc = TestData.random.string
    }

    @Test
    fun getsClefInformation() {
        val expected = ClefResponse(abc)
        every {internalNotationFactory.clef(clefString)} returns clef
        val returnValue = abc
        every { abcService.getAbc(clef) } returns object : AbcProvider {
            override val abc = returnValue
        }
        val actual = clefController.get(clefString)
        assertThat(actual).isEqualTo(expected)
    }
}